### ISSUE TITLE
"Internal error that should never happen" happened

### ISSUE THREAD

**Issue by k15a** [brigand/babel-plugin-flow-react-proptypes#163]
After updating to 13.0.1 from 10.0.0 I've got this error:

```
Module build failed: Error: /Users/konstantin/Development/frontend/projects/bapp/src/components/Theming/types.js: babel-plugin-flow-react-proptypes processing error: This is an internal error that should never happen. Report it immediately with the source file and babel config. Data: {
  "isRequired": false
}
    at makePropType (/Users/konstantin/Development/frontend/node_modules/babel-plugin-flow-react-proptypes/lib/makePropTypesAst.js:341:11)
    at makePropTypesAstForExport (/Users/konstantin/Development/frontend/node_modules/babel-plugin-flow-react-proptypes/lib/makePropTypesAst.js:78:11)
    at PluginPass.ExportNamedDeclaration (/Users/konstantin/Development/frontend/node_modules/babel-plugin-flow-react-proptypes/lib/index.js:381:76)
    at newFn (/Users/konstantin/Development/frontend/projects/bapp/node_modules/babel-core/node_modules/babel-traverse/lib/visitors.js:276:21)
    at NodePath._call (/Users/konstantin/Development/frontend/projects/bapp/node_modules/babel-core/node_modules/babel-traverse/lib/path/context.js:76:18)
    at NodePath.call (/Users/konstantin/Development/frontend/projects/bapp/node_modules/babel-core/node_modules/babel-traverse/lib/path/context.js:48:17)
    at NodePath.visit (/Users/konstantin/Development/frontend/projects/bapp/node_modules/babel-core/node_modules/babel-traverse/lib/path/context.js:105:12)
    at TraversalContext.visitQueue (/Users/konstantin/Development/frontend/projects/bapp/node_modules/babel-core/node_modules/babel-traverse/lib/context.js:150:16)
    at TraversalContext.visitMultiple (/Users/konstantin/Development/frontend/projects/bapp/node_modules/babel-core/node_modules/babel-traverse/lib/context.js:103:17)
    at TraversalContext.visit (/Users/konstantin/Development/frontend/projects/bapp/node_modules/babel-core/node_modules/babel-traverse/lib/context.js:190:19)
    at Function.traverse.node (/Users/konstantin/Development/frontend/projects/bapp/node_modules/babel-core/node_modules/babel-traverse/lib/index.js:114:17)
    at NodePath.visit (/Users/konstantin/Development/frontend/projects/bapp/node_modules/babel-core/node_modules/babel-traverse/lib/path/context.js:115:19)
    at TraversalContext.visitQueue (/Users/konstantin/Development/frontend/projects/bapp/node_modules/babel-core/node_modules/babel-traverse/lib/context.js:150:16)
    at TraversalContext.visitSingle (/Users/konstantin/Development/frontend/projects/bapp/node_modules/babel-core/node_modules/babel-traverse/lib/context.js:108:19)
    at TraversalContext.visit (/Users/konstantin/Development/frontend/projects/bapp/node_modules/babel-core/node_modules/babel-traverse/lib/context.js:192:19)
    at Function.traverse.node (/Users/konstantin/Development/frontend/projects/bapp/node_modules/babel-core/node_modules/babel-traverse/lib/index.js:114:17)

 @ ./src/themes/Elinvar.js 3:58-96
 @ ./src/App.js
 @ ./src/index.js
 @ multi (webpack)-dev-server/client?http://localhost:3000 webpack/hot/dev-server react-dev-utils/webpackHotDevClient ./src/index.js
```

Source file:

```js
// @flow

type $DeepShape<O: Object> = Object &
  $Shape<$ObjMap<O, (<V: Object>(V) => $DeepShape<V>) & (<V>(V) => V)>>;

export type TTheme = {
  colors: {
    background: {
      main: string,
      soft: string,
    },
    seperator: string,
    border: string,
    text: {
      main: string,
      highlight: string,
    },
    table: {
      border: {
        row: string,
        column: string,
      },
    },
    pairs: {
      border: {
        row: string,
        column: string,
      },
    },
    charts: {
      background: string,
      data: [
        {
          major: string,
          minor: string,
        },
        {
          major: string,
          minor: string,
        },
      ],
      spectrum: Array<string>,
      plot: {
        stroke: string,
        fill: string,
      },
      axis: {
        x: {
          line: string,
          ticks: string,
        },
        y: {
          line: string,
          ticks: string,
        },
      },
      tooltip: {
        crosshair: string,
        background: string,
        text: {
          main: string,
          soft: string,
        },
      },
      assetClasses: {
        ALTERNATIVE: string,
        BONDS: string,
        CASH: string,
        EQUITY: string,
        RAW_MATERIALS: string,
      },
    },
  },
};

export type TColors = $PropertyType<TTheme, 'colors'>;
export type TPartialColors = $DeepShape<TColors>;

export type TBaseColors = {
  background: string,
  foreground: string,
  primary: string,
  secondary: string,
  success: string,
  info: string,
  warning: string,
  danger: string,
  // Neutral 1 is near background and neutral 5 is near foreground
  neutral: [string, string, string, string, string],
};

type TBases = {
  light: TBaseColors,
  dark: TBaseColors,
  primary: TBaseColors,
};

export type TBaseName = $Keys<TBases>;

type TSections = {
  body: TBaseName,
  header: TBaseName,
  footer: TBaseName,
  card: TBaseName,
  highlightedCard: TBaseName,
};

export type TSectionName = $Keys<TSections>;

export type TThemeConfig = {
  colors: {
    base: TBases,
    sections: TSections,
    overrides: {
      [ref: TBaseName | TSectionName]: TPartialColors,
    },
  },
};
```

Babel config:

```json
{
  "presets": [
    [
      "env",
      {
        "modules": false
      }
    ],
    "react",
    "flow"
  ],

  "plugins": ["transform-class-properties", "transform-object-rest-spread"],

  "env": {
    "production": {
      "presets": ["react-optimize"]
    },

    "development": {
      "plugins": ["flow-react-proptypes"]
    },

    "test": {
      "plugins": ["transform-es2015-modules-commonjs"]
    }
  }
}
```

If you need more information please let me know.

**Comment by brigand** [comment]
Thanks for the report!

I reduced it down to this:

```js
type $DeepShape<O: Object> = Object &
  $Shape<$ObjMap<O, (<V: Object>(V) => $DeepShape<V>) & (<V>(V) => V)>>;

export type TPartialColors = $DeepShape<TColors>;
```

To be honest, I don't fully understand this code. Could you explain it in english?

I'm not sure if we'll end up fully supporting this in the short term (I have limited time for this project), or bail out and treat it like `any`.

**Comment by k15a** [comment]
[`$Shape`](https://medium.com/netscape/secret-flow-types-86b2ebb30951) is a utility shape from flow:

> Type `$Shape<T>` has two differences from `T`: objects of this type cannot contain properties not described in `T` and may not contain required properties of type `T`.

With `$DeepShape` I have the same behavior as `$Shape` but mapping deeply into the object.

```js
// @flow


type $DeepShape<O: Object> = Object &
  $Shape<$ObjMap<O, (<V: Object>(V) => $DeepShape<V>) & (<V>(V) => V)>>;

type O = {
	a: {
		a1: string,
		a2: string
	},
	b: string
}

const o = {
	a: {
		a1: 'hello'
	}
}

const shape: $Shape<O> = o // This will not work
const deepShape: $DeepShape<O> = o // This will work
```

You can find more context on [`$DeepShape<T>` here](https://github.com/facebook/flow/issues/5262#issuecomment-343803528)

I'am very happy with bailing out and handling this as any.

**Comment by brigand** [comment]
Cool, pushed a fix as 13.0.2.

**Comment by k15a** [comment]
Thank you very much.
